File: programs/sdk/assemblyscript/assembly/receipt.ts
/**
 * Verified receipt facts.
 *
 * A program never sees kernel state. It reads only the facts the core has
 * already verified for a digest the invoking activity named in an explicit
 * grant, decoded from the frozen one hundred and sixteen byte view. The decoded
 * digest is checked against the requested one, so evidence for another receipt
 * cannot be mistaken for evidence for this one.
 */

import { DIGEST_BYTES, IDENTIFIER_BYTES, RECEIPT_ENCODING_BYTES } from "./abi";
import { Amount } from "./amount";
import { equal, pointer, readI32BE, slice } from "./bytes";
import { ERR_EVIDENCE, ERR_RECEIPT_ENCODING, ERR_RESERVED_IDENTIFIER, OK } from "./error";
import { receiptRead } from "./host";
import { ReceiptDigest } from "./ids";

const DIGEST_OFFSET: i32 = 0;
const RESULT_CODE_OFFSET: i32 = 32;
const ASSET_OFFSET: i32 = 36;
const AMOUNT_OFFSET: i32 = 68;
const STATE_ROOT_OFFSET: i32 = 84;

/** Verified receipt facts exposed without raw kernel state. */
export class Receipt {
  digest: StaticArray<u8>;
  resultCode: i32;
  asset: StaticArray<u8>;
  amount: Amount;
  stateRoot: StaticArray<u8>;

  constructor(
    digest: StaticArray<u8>,
    resultCode: i32,
    asset: StaticArray<u8>,
    amount: Amount,
    stateRoot: StaticArray<u8>
  ) {
    this.digest = digest;
    this.resultCode = resultCode;
    this.asset = asset;
    this.amount = amount;
    this.stateRoot = stateRoot;
  }

  /** An empty view, returned alongside a refusal so no fact is invented. */
  static empty(): Receipt {
    return new Receipt(
      new StaticArray<u8>(DIGEST_BYTES),
      0,
      new StaticArray<u8>(IDENTIFIER_BYTES),
      Amount.zero(),
      new StaticArray<u8>(DIGEST_BYTES)
    );
  }

  /** Reports whether the receipt records a successful settlement. */
  settled(): bool {
    return this.resultCode == 0;
  }
}

/** The outcome of one receipt read. */
export class ReceiptRead {
  status: i32;
  receipt: Receipt;

  constructor(status: i32, receipt: Receipt) {
    this.status = status;
    this.receipt = receipt;
  }

  /** Reports whether the host produced verified facts. */
  ok(): bool {
    return this.status == OK;
  }
}

/** Decodes the frozen receipt view the host writes. */
export function decodeReceipt(encoded: StaticArray<u8>): ReceiptRead {
  if (encoded.length != RECEIPT_ENCODING_BYTES) {
    return new ReceiptRead(ERR_RECEIPT_ENCODING, Receipt.empty());
  }
  const receipt = new Receipt(
    slice(encoded, DIGEST_OFFSET, DIGEST_BYTES),
    readI32BE(encoded, RESULT_CODE_OFFSET),
    slice(encoded, ASSET_OFFSET, IDENTIFIER_BYTES),
    Amount.fromBigEndian(encoded, AMOUNT_OFFSET),
    slice(encoded, STATE_ROOT_OFFSET, DIGEST_BYTES)
  );
  return new ReceiptRead(OK, receipt);
}

/** Reads the verified facts of one receipt named by an explicit grant. */
export function readReceipt(receiptDigest: ReceiptDigest): ReceiptRead {
  if (receiptDigest.isReserved()) {
    return new ReceiptRead(ERR_RESERVED_IDENTIFIER, Receipt.empty());
  }
  const encoded = new StaticArray<u8>(RECEIPT_ENCODING_BYTES);
  const outcome = receiptRead(
    pointer(receiptDigest.bytes),
    DIGEST_BYTES,
    pointer(encoded),
    RECEIPT_ENCODING_BYTES
  );
  if (outcome < 0) return new ReceiptRead(outcome, Receipt.empty());
  if (outcome != RECEIPT_ENCODING_BYTES) {
    return new ReceiptRead(ERR_RECEIPT_ENCODING, Receipt.empty());
  }
  const decoded = decodeReceipt(encoded);
  if (!decoded.ok()) return decoded;
  if (!equal(decoded.receipt.digest, 0, receiptDigest.bytes, 0, DIGEST_BYTES)) {
    return new ReceiptRead(ERR_EVIDENCE, Receipt.empty());
  }
  return decoded;
}

File: programs/sdk/rust/src/receipt.rs
//! Verified receipt facts.
//!
//! A program never sees kernel state. It reads only the facts the core has
//! already verified for a digest the invoking activity named in an explicit
//! grant, decoded from the frozen 116-byte view.

use crate::abi::RECEIPT_ENCODING_BYTES;
use crate::amount::Amount;
use crate::error::{Field, ProgramError, Reason};

#[cfg(target_arch = "wasm32")]
use crate::error::HostRefusal;
#[cfg(target_arch = "wasm32")]
use crate::host;
#[cfg(target_arch = "wasm32")]
use crate::ids::ReceiptDigest;

const DIGEST_OFFSET: usize = 0;
const RESULT_CODE_OFFSET: usize = 32;
const ASSET_OFFSET: usize = 36;
const AMOUNT_OFFSET: usize = 68;
const STATE_ROOT_OFFSET: usize = 84;

/// Verified receipt facts exposed without raw kernel state.
#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub struct Receipt {
    /// Digest naming the verified receipt.
    pub digest: [u8; 32],
    /// Result code the receipt records.
    pub result_code: i32,
    /// Asset the receipt settles.
    pub asset: [u8; 32],
    /// Exact integer amount the receipt settles.
    pub amount: Amount,
    /// State root the receipt commits to.
    pub state_root: [u8; 32],
}

impl Receipt {
    /// Decodes the frozen receipt view the host writes.
    ///
    /// # Errors
    ///
    /// Refuses any encoding other than the exact host format.
    pub fn decode(bytes: &[u8]) -> Result<Self, ProgramError> {
        if bytes.len() != RECEIPT_ENCODING_BYTES {
            return Err(ProgramError::value(Field::Receipt, Reason::Malformed));
        }
        Ok(Self {
            digest: fixed::<32>(bytes, DIGEST_OFFSET)?,
            result_code: i32::from_be_bytes(fixed::<4>(bytes, RESULT_CODE_OFFSET)?),
            asset: fixed::<32>(bytes, ASSET_OFFSET)?,
            amount: Amount::from_be_bytes(fixed::<16>(bytes, AMOUNT_OFFSET)?),
            state_root: fixed::<32>(bytes, STATE_ROOT_OFFSET)?,
        })
    }
}

fn fixed<const N: usize>(bytes: &[u8], offset: usize) -> Result<[u8; N], ProgramError> {
    let end = offset
        .checked_add(N)
        .ok_or_else(|| ProgramError::value(Field::Receipt, Reason::Malformed))?;
    let slice = bytes
        .get(offset..end)
        .ok_or_else(|| ProgramError::value(Field::Receipt, Reason::Malformed))?;
    let mut output = [0u8; N];
    output.copy_from_slice(slice);
    Ok(output)
}

/// Reads the verified facts of one receipt named by an explicit grant.
///
/// # Errors
///
/// Refuses missing digest authority, absent or mismatched evidence, and a
/// host encoding outside the frozen view.
#[cfg(target_arch = "wasm32")]
pub fn read(receipt_digest: ReceiptDigest) -> Result<Receipt, ProgramError> {
    let digest = receipt_digest.bytes();
    let mut output = [0u8; RECEIPT_ENCODING_BYTES];
    let status = host::receipt_read(&digest, &mut output)?;
    let written = usize::try_from(status)
        .map_err(|_| ProgramError::value(Field::Receipt, Reason::Malformed))?;
    if written != RECEIPT_ENCODING_BYTES {
        return Err(ProgramError::value(Field::Receipt, Reason::Malformed));
    }
    let receipt = Receipt::decode(&output)?;
    if receipt.digest != digest {
        return Err(ProgramError::Host(HostRefusal::Evidence));
    }
    Ok(receipt)
}
